README.md

```diff
@@ -198,289 +189,492 @@ e.increment();  // logs 1
+  }
+  set(obj, value) {
+    return this.#set(obj, value);
+  }
 }
+```
+
+This example could be roughly desugared as follows:
 
-export decorator @callable {
-  @call(function(...args) { return this.call(...args); })
+```mjs
+let initialize, get, set;
+export class Box {
+  #_contents = initialize(undefined);
+  get #contents() { return get.call(this); }
+  set #contents(v) { set.call(this, v); }
+
+  static {
+    get = function() { return this.#_contents; },
+    set = function(v) { this.#_contents = v; }
+  }
 }
+({get, set, initialize} = key.show({get, set}, {kind: "field", isStatic: false}));
 ```
 
-Note that a decorator like `@call` could be considered for a future built-in decorator, in a way that avoids creating an additional subclass.
-
-### Combined example
+### `@deprecated`
 
-Some of the above examples could be combined to form a mini-framework, to make it easier to write HTML Custom Elements.
+The `@deprecated` decorator prints warnings when a deprecated field, method or accessor is used. As an example usage:
 
 ```mjs
-import { @set } from "./set.mjs";
-import { @tracked } from "./tracked.mjs";
-import { @bound } from "./bound.mjs";
-import { @defineElement } from "./defineElement.mjs";
+import { deprecated } from "./deprecated.mjs"
 
-@defineElement('counter-widget')
-class CounterWidget extends HTMLElement {
-  @tracked x = 0;
+export class MyClass {
+  @deprecated field;
+
+  @deprecated method() { }
+
+  otherMethod() { }
+}
+```
 
-  @set onclick = this.clicked;
+To allow the `deprecated` to work on different kinds of class elements, the `kind` field of the context object lets decorators see which kind of syntactic construct they are deprecating. This technique also allows an error to be thrown when the decorator is used in a context where it can't apply--for example, the entire class cannot be marked as deprecated, since there is no way to intercept its access.
 
-  @bound clicked() { this.x++; }
+```mjs
+// deprecated.mjs
 
-  connectedCallback() { this.render(); }
+function wrapDeprecated(fn) {
+  let name = fn.name
+  function method(...args) {
+    console.warn(`call to deprecated code ${name}`);
+    return fn.call(this, ...args);
+  }
+  method.name = name;
+  return method;
+}
 
-  render() { this.textContent = this.x.toString(); }
+export function deprecated(element, {kind}) {
+  switch (kind) {
+    case 'method':
+    case 'getter':
+    case 'setter':
+      return wrapDecorated(element);
+    case 'field': {
+      let { get, set } = element;
+      return { get: wrapDeprecated(get), set: wrapDeprecated(set) };
+    }
+    default: // includes 'class'
+      throw new Error(`Unsupported @deprecated target ${kind}`);
+  }
 }
 ```
 
-## The idea
+The desugaring here is analogous to the above examples, which show the use of `kind`.
 
-A new goal of this proposal: It should be possible for tools and JS engines to understand what's going on with decorators.
+## Method decorators requiring initialization work
```

Comment:
Could it be an option to call all field / method decorators another time, with a special flag indicating the phase, e.g. `options.phase` is `"declaration"` | `"initialization"`and in case of `initialization` phase the `this` instance available as `option.instance`? This does add overhead to the class construction, but I can imagine this is in practice optimisable as JITs might have an idea which decorators are noops during initialization.

Reply:
This proposal is based on the design goal that it be optimizable not just in JITs but also in interpreters. So I would rather maintain separate syntax for it. Many people are asking the same question here, so it sounds like I should clarify the explanation further.